library/Fission/Storage/IPFS/Add.hs

```diff
@@ -0,0 +1,167 @@
+import           Fission.IPFS.DAG.Link       as DAG.Link
+
+import           Fission.Storage.IPFS.DAG as DAG
+import qualified Fission.Storage.IPFS.Pin as IPFS.Pin
+
+addRaw ::
+  ( MonadRIO          cfg m
+  , HasProcessContext cfg
+  , HasLogFunc        cfg
+  , Has HTTP.Manager  cfg
+  , Has IPFS.URL      cfg
+  , Has IPFS.BinPath  cfg
+  , Has IPFS.Timeout  cfg
+  )
+  => Lazy.ByteString
+  -> m (Either IPFS.Error.Add IPFS.CID)
+addRaw raw =
+  IPFS.Proc.run ["add", "-HQ"] raw >>= \case
+    (ExitSuccess, result, _) ->
+      case CL.lines result of
+        [cid] ->
+          cid
+            |> UTF8.textShow
+            |> UTF8.stripN 1
+            |> mkCID
+            |> IPFS.Pin.add
+
+        bad ->
+          return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+addFile ::
+  ( MonadRIO          cfg m
+  , HasProcessContext cfg
+  , HasLogFunc        cfg
+  , Has HTTP.Manager  cfg
+  , Has IPFS.URL      cfg
+  , Has IPFS.BinPath  cfg
+  , Has IPFS.Timeout  cfg
+  )
+  => Lazy.ByteString
+  -> IPFS.Name
+  -> m (Either IPFS.Error.Add IPFS.SparseTree)
+addFile raw name =
+  IPFS.Proc.run opts raw >>= \case
+    (ExitSuccess, result, _) ->
+      IPFS.Pin.add (mkCID <| UTF8.textShow result) >>= \case
+        Left err ->
+          return . Left . UnknownAddErr <| UTF8.textShow err
+
+        Right _ ->
+          case CL.lines result of
+            [inner, outer] ->
+              let
+                sparseTree  = Directory [(Hash rootCID, fileWrapper)]
+                fileWrapper = Directory [(fileName, Content fileCID)]
+                rootCID     = CID <| UTF8.textShow outer
+                fileCID     = CID . UTF8.stripN 1 <| UTF8.textShow inner
+                fileName    = Key name
+              in
+                return <| Right sparseTree
+
+            bad ->
+              return . Left . UnexpectedOutput <| UTF8.textShow bad
+
+
+    (ExitFailure _, _, err) ->
+      return . Left . UnknownAddErr <| UTF8.textShow err
+
+    where
+      opts = [ "add"
+             , "-wq"
+             , "--stdin-name"
+             , unName name
+             ]
+
+addPath ::
+  ( RIOProc           cfg m
+  , Has IPFS.Timeout  cfg
+  , Has IPFS.BinPath  cfg
+  )
+  => FilePath
+  -> m (Either IPFS.Error.Add CID)
+addPath path = IPFS.Proc.run ["add", "-HQ", path] "" >>= pure . \case
+    (ExitSuccess, result, _) ->
+      case CL.lines result of
+        [cid] -> Right . mkCID . UTF8.stripN 1 <| UTF8.textShow cid
+        bad   -> Left . UnexpectedOutput <| UTF8.textShow bad
+
+    (ExitFailure _, _, err) ->
+      Left . UnknownAddErr <| UTF8.textShow err
+
+addDir ::
+  ( MonadRIO cfg m
+  , HasProcessContext cfg
+  , HasLogFunc cfg
+  , Has IPFS.BinPath cfg
+  , Has IPFS.Timeout cfg
+  )
+  => [Glob.Pattern]
+  -> FilePath
+  -> m (Either IPFS.Error.Add IPFS.CID)
+addDir ignored path = doesFileExist path >>= \case
```

Comment:
removed the old `addDir` is there any time when we'd want to do this natively? I tested it out and got the exact same hashes natively & through this

Reply:
I can't see this commit anymore apparently, but from memory: we don't do it natively because this module assumes that we're on the EC2 instance, and actual storage happens on the S3 node.

Reply:
ah sorry, I think "natively" was the wrong word to use here. I just meant our recurse vs IPFS' recurse. In other words, manually creating DAG nodes vs `ipfs add -r ...`

Neither one should affect that datastores on the back end. And `addDir` is never actually called in the Web API, only the CLI. This is because files to be added through the API, are serialized and added as raw data. Directories are never crawled.

Reply:
>  And addDir is never actually called in the Web API, only the CLI

We really need to split this into its own library. It's starting to get confusing.